jetstream/analysis.py

```diff
@@ -457,3 +446,47 @@ def run(self, current_date: datetime, dry_run: bool = False) -> None:
 
         result_futures = client.compute(results)
         client.gather(result_futures)  # block until futures have finished
+
+    def ensure_enrollments(
+        self, current_date: datetime, recreate_enrollments: bool = False
+    ) -> None:
+        """Ensure that enrollment tables for experiment are up-to-date or re-create."""
+        time_limits = self._get_timelimits_if_ready(AnalysisPeriod.DAY, current_date)
+
+        if time_limits is None:
+            logger.info("Skipping %s (%s); not ready", self.config.experiment.normandy_slug)
+            return
+
+        if self.config.experiment.start_date is None:
+            raise errors.NoStartDateException(self.config.experiment.normandy_slug)
+
+        normalized_slug = bq_normalize_name(self.config.experiment.normandy_slug)
+        enrollments_table = f"enrollments_{normalized_slug}"
+
+        try:
+            if not recreate_enrollments:
+                # check if enrollments table already exists and skip creation
+                try:
+                    self.bigquery.client.get_table(enrollments_table)
```

Comment:
Yeah, I was looking at the docs here: https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.job.WriteDisposition.html

It says _Creation, truncation and append actions occur as one atomic update upon job completion._ So wouldn't that mean the query is still executed (and billed) but then would fail because the table already exists?

Reply:
I don't think that implies that jobs always execute -- just that if it does, consumers always see a consistent view.

https://colab.research.google.com/drive/1OJFlmNvIWi7Yo29QawRrik3v3vGmaP5F#scrollTo=6BodY7sqX-Y4 shows that if we try to run a query twice with the same destination table, caching disabled, and disposition WRITE_EMPTY, the second attempt will return immediately, so it doesn't seem like it's doing any work.

Reply:
Interesting. In that case, this could work.